### PR TITLE
fix(kuma-cp) use cmux for MADS server

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

The existing approach of multiplexing GRPC and HTTP/1.1 on the same port doesn't work without TLS. Instead, library [cmux](https://github.com/soheilhy/cmux) works pretty well, that's why current PR replaces the existing approach with `cmux`.

Change default version of MADS API to v1alpha1 because of https://github.com/kumahq/kuma/issues/1886

### Full changelog

* add new library `cmux`
* unit tests for MADS server
* change default version of MADS API to `v1alpha1`

### CODE REVIEW

File: pkg/mads/server/server.go [hunk 1]
```diff
@@ -117,34 +100,82 @@ func (s *muxServer) createHttpServicesHandler() http.Handler {
...
-			grpcServer:  s.createGRPCServer(),
-			passthrough: s.createHttpServicesHandler(),
-		},
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", s.config.Port))
+	if err != nil {
+		return err
 	}
+	defer func() {
+		if err := l.Close(); err != nil {
+			log.Error(err, "")
```
Comment (jakubdyszkiewicz) [inline pkg/mads/server/server.go:109]: please provide an error message

File: pkg/mads/server/server.go [hunk 2]
```diff
@@ -117,34 +100,82 @@ func (s *muxServer) createHttpServicesHandler() http.Handler {
...
 		defer close(errChan)
-		err := server.ListenAndServe()
+		err := m.Serve()
 		if err != nil {
 			switch err {
 			case http.ErrServerClosed:
 				log.Info("shutting down server")
 			default:
-				log.Error(err, "could not start an HTTP Server")
+				log.Error(err, "could not start an mux Server")
```
Comment (jakubdyszkiewicz) [inline pkg/mads/server/server.go:161]: nit: `a mux Server`

File: pkg/mads/server/server_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	BeforeEach(func() {
+		m, err := metrics.NewMetrics("zone-1")
+		Expect(err).ToNot(HaveOccurred())
+
+		rt = &testRuntime{
+			rm:      manager.NewResourceManager(memory.NewStore()),
+			config:  kuma_cp.Config{MonitoringAssignmentServer: mads.DefaultMonitoringAssignmentServerConfig()},
+			metrics: m,
+		}
+		rt.config.MonitoringAssignmentServer.Port = 15676
```
Comment (jakubdyszkiewicz) [inline pkg/mads/server/server_test.go:65]: no port hardcoding in tests! please randomize the port

File: pkg/mads/server/server_test.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+		err = stream.RequestAssignments("client-1")
+		Expect(err).ToNot(HaveOccurred())
+
+		assignments, err := stream.WaitForAssignments()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(assignments).To(HaveLen(0))
+	})
+
+	It("", func() {
```
Comment (jakubdyszkiewicz) [inline pkg/mads/server/server_test.go:101]: it should...

File: pkg/plugins/ca/builtin/config/builtin_ca_config.pb.go [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.23.0
+// 	protoc-gen-go v1.25.0
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/ca/builtin/config/builtin_ca_config.pb.go:3]: upgrade go and do `make dev/tools`

**Review by jakubdyszkiewicz — COMMENTED** [review]
nice! 👏

File: pkg/plugins/ca/builtin/config/builtin_ca_config.pb.go (hunk 5 above)
Comment (lobkovilya) [reply pkg/plugins/ca/builtin/config/builtin_ca_config.pb.go]: Thanks! That was extremely helpful, I had the latest Go version, but didn't call `make dev/tools`

File: pkg/mads/server/server_test.go (hunk 3 above)
Comment (lobkovilya) [reply pkg/mads/server/server_test.go]: Replaced with`PickTCPPort`, is it okay?

**Review by nickolaev — APPROVED** [review]

File: pkg/mads/server/server_test.go (hunk 3 above)
Comment (jakubdyszkiewicz) [reply pkg/mads/server/server_test.go]: it is

**Review by jakubdyszkiewicz — APPROVED** [review]

File: pkg/mads/server/server_test.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
+package server_test
```
Comment (austince) [inline pkg/mads/server/server_test.go:1]: Thanks for this nice test!

**Review by austince — APPROVED** [review]

**Comment by austince** [comment]
Should we add the same to the dp-server @lobkovilya? 
https://github.com/kumahq/kuma/blob/10936f8e8b87f5681eae2187aeffed6fb5b80774/pkg/dp-server/server/server.go#L104-L110

Or is this always run with TLS so it doesn't matter?

**Pull request merged**